tree.go

```diff
@@ -22,33 +21,34 @@ type Tree interface {
 	Root() ([]byte, error)
 }
 
-var _ Tree = &DefaultTree{}
+// trees is a global map used for keeping track of registered tree constructors for JSON serialization
+// The keys of this map should be kebab cased. E.g. "default-tree"
+var trees = make(map[string]TreeConstructorFn)
+
+func RegisterTree(treeName string, treeConstructor TreeConstructorFn) error {
```

Comment:
Let's document and add a note that it must be called in `init` if you don't want to synchronize the map.

Reply:
The safest option, though, is to still synchronize the map.

Reply:
We don't do it for codecs because it's not allowed to register user-defined codecs

Reply:
i switch using `sync.map` as requested